bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/container/ModuleCapability.java

```diff
@@ -93,4 +96,26 @@ public ModuleRevision getResource() {
 	public String toString() {
 		return namespace + ModuleContainer.toString(attributes, false) + ModuleContainer.toString(directives, true);
 	}
+
+	@Override
+	public int hashCode() {
+		if (hashCode != 0) {
+			return hashCode;
+		}
+		return hashCode = Objects.hash(namespace, directives, attributes, revision);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj) {
+			return true;
+		}
+		if (obj instanceof Capability) {
+			Capability other = (Capability) obj;
+			return Objects.equals(namespace, other.getNamespace()) && Objects.equals(directives, other.getDirectives())
+					&& Objects.equals(attributes, other.getAttributes())
+					&& Objects.equals(revision, other.getResource());
```

Comment:
because the `revision` here is always a `BundleRevision` the `Objects.equals` call is always going to return false here unless the `obj` is somehow another `Capability` from the same `BundleRevision`.

Reply:
But in that case it would have to be a case where a `BundleRevision has two capabilities with the exact same attributes and directives.

Reply:
>  unless the obj is somehow another Capability from the same BundleRevision.

This is only true when you assume that `obj` can only ever come from the running equinox system. But assume for example the class `aQute.bnd.osgi.resource.CapabilityImpl` is used and I construct it from a ModuleCapability:

```
ModuleCapability moduleCapability = ....;
aQute.bnd.osgi.resource.CapabilityImpl wrapper = new CapabilityImpl(moduleCapability.getNamespace(),
		moduleCapability.getResource(), moduleCapability.getDirectives(), moduleCapability.getAttributes());
System.out.println(moduleCapability.equals(wrapper)); // --> must be true
System.out.println(wrapper.equals(moduleCapability)); // --> must be true
```

then I have two `Capability` objects that should per specification compare to true when calling equals and produce the same hashcode. I hope this makes it more clear what this PR is about

Thinking more about it for performance reason one might compare the resource first, but that's more an implementation decision.

Reply:
I see, but seems strange that they wrap the capability but not the resource.

I definitely would want the resource equality checked before checking the attrs/directives.

Reply:
> I definitely would want the resource equality checked before checking the attrs/directives.

Done!